### DISCUSSION TITLE
Generally, how are library summaries created?

### DISCUSSION THREAD

**Discussion by khatchad** [wala/WALA#1334] [Q&A] [answered]
Looking at [`com.ibm.wala.ipa.summaries`](https://wala.github.io/javadoc/com/ibm/wala/ipa/summaries/package-summary.html), I see that it is possible to summarize the analysis results of libraries using an XML file. But, how are these summaries generally created? Are they created by hand? Is there any documentation on how to create library summaries?

**Comment by msridhar** [comment, marked as answer]
I believe the XML summaries were created (long ago!) by hand.  No documentation on how to create them, unfortunately.  But I hope that the XML format isn't too complex, so one could figure out how to modify / extend if needed.

Note that I believe these summaries are only relevant to analysis of Java code; I'm not sure if there is a pipeline set up to load similar summaries for other languages.

**Comment by khatchad** [comment]
Thanks, @msridhar. But, when analyzing Java code using WALA, the JDK is included (sans the "exclusions" file), right? Are these summaries mainly for third-party libraries? It would be useful to have a tool that creates them.

I'm unsure about Javascript, but they seem to be extensively used for Python; @tatianacv can confirm.

**Reply by tatianacv** [reply]
Yes, they are used extensively for Python to model ML frameworks.
